src/Renci.SshNet/SshCommand.cs

```diff
@@ -356,13 +357,12 @@ public string EndExecute(IAsyncResult asyncResult)
         /// <summary>
         /// Cancels command execution in asynchronous scenarios.
         /// </summary>
-        public void CancelAsync()
+        /// <param name="forceKill">if true send SIGKILL instead of SIGTERM.</param>
+        public void CancelAsync(bool forceKill = false)
         {
-            if (_channel is not null && _channel.IsOpen && _asyncResult is not null)
-            {
-                // TODO: check with Oleg if we shouldn't dispose the channel and uninitialize it ?
-                _channel.Dispose();
-            }
+            var signal = forceKill ? "KILL" : "TERM";
+            _ = _channel?.SendExitSignalRequest(signal, coreDumped: false, "Command execution has been cancelled.", "en");
+            _ = _commandCancelledWaitHandle.Set();
```

Comment:
Sure. I am doing some further debugging. It seem that we might have to also execute the ```_callback``` as well as Flushing the ```OutputStream``` + ```ExntednedOutputStream``` .

On my local testing, ```Channel.Close``` event is never called and when I try to make ```SendExitSignalRequest``` wait for a response, I got ```Session Operation has Timeout``` Exception. Wondering if anyone else can recreate it.

Reply:
Is it because of the call to `UnsubscribeFromEventsAndDisposeChannel(_channel);` in `EndExecute`?

Perhaps `_commandCancelledWaitHandle` should be removed? If I read https://datatracker.ietf.org/doc/html/rfc4254#section-6.10 correctly, we should always get SSH_MSG_CHANNEL_CLOSE after the signal, so it "should" just work.

Reply:
Yeah, I was expecting it to just work. However, when I tried removing the _commandCancelledWaitHandle and made SendExitSignalRequest wait for a response, it never happened. If I just make SendExitSignalRequest send the message without waiting for a response, it will be blocked until the command actually finishes. 
My server openSSH version: OpenSSH_8.2p1 Ubuntu-4ubuntu0.11 so those signal is probably already supported.
I will takes a deeper look at the log.